persist/fs/commitlog/commit_log.go

```diff
@@ -0,0 +1,386 @@
+
+	// TODO(r): parameterize commit log write retries and retry delay in future
+	writeRetries    = 2
+	writeRetryDelay = time.Millisecond
+
+	timeZero = time.Time{}
+)
+
+type completionFn func(err error)
+
+type commitLog struct {
+	sync.RWMutex
+	opts    m3db.DatabaseOptions
+	metrics xmetrics.Scope
+	nowFn   m3db.NowFn
+
+	writer commitLogWriter
+
+	// TODO(r): replace buffered channel with concurrent striped
+	// circular buffer to avoid central write lock contention
+	writes chan commitLogWrite
+
+	flushMutex      sync.RWMutex
+	lastFlushAt     time.Time
+	pendingFlushFns []completionFn
+
+	bitset         bitset
+	writerExpireAt time.Time
+	closed         bool
+	closeErr       chan error
+}
+
+type valueType int
+
+const (
+	writeValueType valueType = iota
+	flushValueType
+)
+
+type commitLogWrite struct {
+	valueType valueType
+
+	series       m3db.CommitLogSeries
+	datapoint    m3db.Datapoint
+	unit         xtime.Unit
+	annotation   m3db.Annotation
+	completionFn completionFn
+}
+
+// NewCommitLog creates a new commit log
+func NewCommitLog(opts m3db.DatabaseOptions) (m3db.CommitLog, error) {
+	l := &commitLog{
+		opts:     opts,
+		nowFn:    opts.GetNowFn(),
+		writes:   make(chan commitLogWrite, opts.GetCommitLogBacklogQueueSize()),
+		metrics:  opts.GetMetricsScope().SubScope("commitlog"),
+		closeErr: make(chan error),
+	}
+
+	// Open the buffered commit log writer
+	if err := l.openWriter(l.nowFn()); err != nil {
+		return nil, err
+	}
+
+	// Flush the info header to ensure we can write to disk
+	if err := l.writer.Flush(); err != nil {
+		return nil, err
+	}
+
+	// Asynchronously write
+	go l.write()
+
+	flushInterval := l.opts.GetCommitLogFlushInterval()
+	if flushInterval > 0 {
+		// Continually flush the commit log at given interval if set
+		go l.flushEvery(flushInterval)
+	}
+
+	return l, nil
+}
+
+func (l *commitLog) flushEvery(interval time.Duration) {
+	// Periodically flush the underlying commit log writer to cover
+	// the case when writes stall for a considerable time
+	var sleepForOverride time.Duration
+	for {
+		sleepFor := interval
+		if sleepForOverride > 0 {
+			sleepFor = sleepForOverride
+			sleepForOverride = 0
+		}
+
+		time.Sleep(sleepFor)
+
+		l.flushMutex.RLock()
+		lastFlushAt := l.lastFlushAt
+		l.flushMutex.RUnlock()
+
+		sinceLastFlush := l.nowFn().Sub(lastFlushAt)
+		if sinceLastFlush < interval {
+			// Flushed already recently, sleep until we would next consider flushing
+			sleepForOverride = interval - sinceLastFlush
+			continue
+		}
+
+		// Request a flush
+		l.RLock()
+		if l.closed {
+			l.RUnlock()
+			return
+		}
+		l.writes <- commitLogWrite{valueType: flushValueType}
+		l.RUnlock()
+	}
+}
+
+func (l *commitLog) write() {
+	var (
+		log             = l.opts.GetLogger()
+		write, retry    commitLogWrite
+		retryInProgress bool
+		retryAttempts   int
+		err             error
+		open            bool
+	)
+	for {
+		if retryInProgress && retryAttempts < 0 {
```

Comment:
Looks like when we decide to retry, we set `retryAttempts` to `writeRetries`. In that case, seems we should do
```
if retryInProgress && retryAttempts > 0 {
  retryAttempts--
  ...
}
```
?